src/index.cpp

```diff
@@ -3007,119 +3100,130 @@ template <typename T, typename TagT, typename LabelT> void Index<T, TagT, LabelT
     delete[] bfs_sets;
 }
 
+// REFACTOR: This should be an OptimizedDataStore class, dummy impl here for compiling sake
 template <typename T, typename TagT, typename LabelT> void Index<T, TagT, LabelT>::optimize_index_layout()
 { // use after build or load
-    if (_dynamic_index)
-    {
-        throw diskann::ANNException("Optimize_index_layout not implemented for dyanmic indices", -1, __FUNCSIG__,
-                                    __FILE__, __LINE__);
-    }
-
-    _data_len = (_aligned_dim + 1) * sizeof(float);
-    _neighbor_len = (_max_observed_degree + 1) * sizeof(uint32_t);
-    _node_size = _data_len + _neighbor_len;
-    _opt_graph = new char[_node_size * _nd];
-    DistanceFastL2<T> *dist_fast = (DistanceFastL2<T> *)_distance;
-    for (uint32_t i = 0; i < _nd; i++)
-    {
-        char *cur_node_offset = _opt_graph + i * _node_size;
-        float cur_norm = dist_fast->norm(_data + i * _aligned_dim, _aligned_dim);
-        std::memcpy(cur_node_offset, &cur_norm, sizeof(float));
-        std::memcpy(cur_node_offset + sizeof(float), _data + i * _aligned_dim, _data_len - sizeof(float));
-
-        cur_node_offset += _data_len;
-        uint32_t k = _final_graph[i].size();
-        std::memcpy(cur_node_offset, &k, sizeof(uint32_t));
-        std::memcpy(cur_node_offset + sizeof(uint32_t), _final_graph[i].data(), k * sizeof(uint32_t));
-        std::vector<uint32_t>().swap(_final_graph[i]);
-    }
-    _final_graph.clear();
-    _final_graph.shrink_to_fit();
 }
 
+// REFACTOR: This should be an OptimizedDataStore class
```

Comment:
Ok, we wanted another PR, since optimized index layout needs _data and _graph in same malloc, so that would actually be a joint data and graph in-mem store.. 

Let me add back some code to get this working with our new data store